lib/ios-deploy.js

```diff
@@ -32,15 +28,75 @@ class IOSDeploy {
   }
 
   async install (app) {
-    const args = [`--id`, this.udid, `--bundle`, app];
     try {
-      await retryInterval(5, 500, exec, this.cmd, args);
+      const bundlePathOnPhone = await this.pushAppBundle(app);
+      await this.installApplcation(bundlePathOnPhone);
     } catch (err) {
-      logger.debug(`Stdout: '${err.stdout}'. Stderr: '${err.stderr}'.`);
       throw new Error(`Could not install app: '${err.message}'`);
     }
   }
 
+  async installApplcation (bundlePathOnPhone) {
+    const notificationService = await services.startNotificationProxyService(this.udid);
+    const installationService = await services.startInstallationProxyService(this.udid);
+    const appInstalledNotification = new B((resolve) => {
+      notificationService.observeNotification(APPLICATION_INSTALLED_NOTIFICATION, { notification: resolve });
+    });
+    try {
+      await installationService.installApplication(bundlePathOnPhone, { PackageType: 'Developer'});
+      await appInstalledNotification;
+    } finally {
+      installationService.close();
+      notificationService.close();
+    }
+  }
+
+  async pushAppBundle (app) {
+    const afcService = await services.startAfcService(this.udid);
+    try {
+      const bundlePathOnPhone = await this.createAppPath(afcService, app);
+      const promises = [];
+      await this.walkDir(app, async (itemPath, isDir) => {
+        const pathOnPhone = path.join(bundlePathOnPhone, path.relative(app, itemPath));
+        if (isDir) {
+          await afcService.createDirectory(pathOnPhone);
+        } else {
+          const readStream = fs.createReadStream(itemPath, {autoClose: true});
+          const writeStream = await afcService.createWriteStream(pathOnPhone, {autoDestroy: true });
+          promises.push(new B((resolve) => writeStream.on('close', resolve)));
+          readStream.pipe(writeStream);
+        }
+      });
+      await B.all(promises).timeout(APPLICATION_PUSH_TIMEOUT);
+      return bundlePathOnPhone;
+    } finally {
+      afcService.close();
+    }
+  }
+
+  async createAppPath (afcService, localAppPath) {
+    const basename = path.basename(localAppPath);
+    const relativePath = path.join(INSTALLATION_STAGING_DIR, basename);
+    try {
+      await afcService.deleteDirectory(relativePath);
+    } catch (ign) {}
+    await afcService.createDirectory(relativePath);
+    return relativePath;
+  }
+
+  async walkDir (dir, callback) {
```

Comment:
This can be made simpler:
```js
async walkDir (dir, callback) {
  for (const file of await fs.readdir(dir, {withFileTypes: true})) {
    const itemPath = path.join(dir, file.name);
    const isDirectory = file.isDirectory();
    await callback(itemPath, isDirectory);
    if (isDirectory) {
      await this.walkDir(itemPath, callback);
    }
  }
}
```
Alternatively, a glob can be used:

```js
async walkDir (dir, callback) {
  for (const itemPath of await fs.glob(`${dir}/**`)) {
    const isDirectory = (await fs.stat(itemPath)).isDirectory();
    await callback(itemPath, isDirectory);
  }
}
```

Reply:
Good idea. I'm using the first one